services/debugger/transformation/transformationStatusUploader.go

```diff
@@ -295,3 +224,122 @@ func getEventsAfterTransform(singularEvent types.SingularEventT, receivedAt time
 		EventPayloads: []*EventPayloadAfterTransform{getEventAfterTransform(singularEvent)},
 	}
 }
+
+func recordHistoricTransformations(tIDs []string) {
+	for _, tID := range tIDs {
+		var tStatuses []*TransformationStatusT
+		transformationCacheMapLock.Lock()
+		if _, ok := transformationCacheMap[tID]; ok {
+			tStatuses = transformationCacheMap[tID]
+			delete(transformationCacheMap, tID)
+		}
+		transformationCacheMapLock.Unlock()
+		for _, tStatus := range tStatuses {
+			processRecordTransformationStatus(tStatus, tID)
+		}
+	}
+}
+
+func populateEventsCache(cache map[string][]*TransformationStatusT, transformationID string,
+	transformation *TransformationStatusT) {
+	transformationCacheMapLock.Lock()
+	defer transformationCacheMapLock.Unlock()
+
+	if _, ok := cache[transformationID]; !ok {
+		cache[transformationID] = make([]*TransformationStatusT, 0, maxTransformationCacheSize)
+	}
+	eventsCache := cache[transformationID]
+	eventsCache = append(eventsCache, transformation)
+	if len(eventsCache) > maxTransformationCacheSize {
+		eventsCache = eventsCache[len(eventsCache)-maxTransformationCacheSize:]
+	}
+	cache[transformationID] = eventsCache
+}
+
+func processRecordTransformationStatus(tStatus *TransformationStatusT, tID string) {
+	reportedMessageIDs := make(map[string]struct{})
+	eventBeforeMap := make(map[string]*EventBeforeTransform)
+	eventAfterMap := make(map[string]*EventsAfterTransform)
+	for _, userTransformerEvent := range tStatus.UserTransformedEvents {
+		if userTransformerEvent.Metadata.MessageID != "" {
+			reportedMessageIDs[userTransformerEvent.Metadata.MessageID] = struct{}{}
+			singularEventWithReceivedAt := tStatus.EventsByMessageID[userTransformerEvent.Metadata.MessageID]
+			if _, ok := eventBeforeMap[userTransformerEvent.Metadata.MessageID]; !ok {
+				eventBeforeMap[userTransformerEvent.Metadata.MessageID] = getEventBeforeTransform(singularEventWithReceivedAt.SingularEvent, singularEventWithReceivedAt.ReceivedAt)
+			}
```

Comment:
This section of code, is rather hard to read. What if:
```suggestion
                metadata := userTransformerEvent.Metadata
		if metadata.MessageID != "" {
			reportedMessageIDs[metadata.MessageID] = struct{}{}
			singularEventWithReceivedAt := tStatus.EventsByMessageID[metadata.MessageID]
			if _, ok := eventBeforeMap[metadata.MessageID]; !ok {
				eventBeforeMap[metadata.MessageID] = getEventBeforeTransform(singularEventWithReceivedAt.SingularEvent, singularEventWithReceivedAt.ReceivedAt)
			}
```

Reply:
Introduce variables for the object parameters which are being referenced multiple time